cmake/modules/boards.cmake

```diff
@@ -0,0 +1,189 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# Validate board and setup boards target.
+#
+# This CMake module will validate the BOARD argument as well as splitting the
+# BOARD argument into <BOARD> and <BOARD_REVISION>.
+#
+# If a board implementation is not found for the specified board an error will
+# be raised and list of valid boards will be printed.
+#
+# If user provided board is a board alias, the board will be adjusted to real
+# board name.
+#
+# If board name is deprecated, then board will be adjusted to new board name and
+# a deprecation warning will be printed to the user.
+#
+# Outcome:
+# The following variables will be defined when this CMake module completes:
+#
+# - BOARD:          Board, without revision field.
+# - BOARD_REVISION: Board revision
+# - BOARD_DIR:      Board directory with the implementation for selected board
+# - ARCH_DIR:       Arch dir for extracted from selected board
+#
+# The following targets will be defined when this CMake module completes:
+# - board  : when invoked a list of valid boards will be printed
+#
+# Variable dependencies:
+# - BOARD_ROOT: CMake list of board roots containing board implementations
+# - ARCH_ROOT:  CMake list of arch roots containing arch implementations
+#
+# CMake module dependencies:
+# - extension CMake module.
+
+include_guard(GLOBAL)
+
+# Dependencies of this module.
+include(extensions)
+
+# This function will print a list of available boards.
+# It can be executed during CMake configure time or script mode, or be
+# defined as a build target by setting the `TARGET` flag when called.
+#
+# Arguments:
+# TARGET:   Flag to specify that build target `board` should be created
+# FILE_OUT: Output file to write instead of standard out. This argument is
+#           ignored if TARGET is specified.
+function(usage_boards)
+  cmake_parse_arguments(BOARD "TARGET" "FILE_OUT" "" ${ARGN})
+
+  set(arch_root_args)
+  foreach(root ${ARCH_ROOT})
+    list(APPEND arch_root_args "--arch-root=${root}")
+  endforeach()
+
+  set(board_root_args)
+  foreach(root ${BOARD_ROOT})
+    list(APPEND board_root_args "--board-root=${root}")
+  endforeach()
+
+  set(list_boards_commands
+    COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/list_boards.py
+            ${arch_root_args} ${board_root_args}
+  )
+
+  if(BOARD_TARGET)
+    add_custom_target(boards ${list_boards_commands} USES_TERMINAL)
+  else()
+    if (DEFINED BOARD_FILE_OUT)
+      list(APPEND list_boards_commands OUTPUT_FILE "${BOARD_FILE_OUT}")
+    endif()
+
+    execute_process(${list_boards_commands})
+  endif()
+endfunction()
+
+if(CMAKE_SCRIPT_MODE_FILE)
+  # Called in script mode. No board validation, just return.
+  return()
+endif()
```

Comment:
on main, shields.cmake is invoked in script mode here:
https://github.com/zephyrproject-rtos/zephyr/blob/7f5eb921dc8bb1b7014886986bbffdf111b68b15/cmake/app/boilerplate.cmake#L403
but as you have noticed this has changed into a direct message printed with this PR.

`boards.cmake` and `shields.cmake` did use to support script mode, so I prefer to keep that support because if we remove it then the proper way would be to deprecate script mode before removal.

There may be external users who uses those scripts to fetch known boards / shields in the build system.
Of course they will need to adopt to new location, but the functionality is still intact.

Secondly, I have received unofficial question whether we could also print valid revisions for boards that supports multiple revisions. As the revisions are known inside CMake, but not directly known in the python script, then the question is also if we should rethink board printing.

Duplicate revision logic in python script https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/list_boards.py or obtain board revision knowledge through the CMake build system.
With the latter, the PR #41302 may become useful.

Reply:
IMHO we should remove the support for executing boards.cmake and shields.cmake.

It is not documented that it is supported, so technically, it is not officially supported IMHO. It's just something we did internally.

We now have a different mechanism for sharing CMake code with external systems. These modules. If we want to retain the functionality that one had by executing boards.cmake and shields.cmake then that should be done through modules, not by supporting third-parties executing our .cmake files.

But if you still believe that it should be supported, then it should be documented what happens when boards.cmake and shields.cmake is executed in script mode.

Reply:
> It is not documented that it is supported, so technically, it is not officially supported IMHO. It's just something we did internally.
> 

Most of the CMake stuff we have do is not properly documented :disappointed: 
So hard to judge whether something should be considered externally used or not.

> We now have a different mechanism for sharing CMake code with external systems. These modules. If we want to retain the functionality that one had by executing boards.cmake and shields.cmake then that should be done through modules, not by supporting third-parties executing our .cmake files.
> 

Which is what this PR is about: #41302 
a uniform way we can invoke just subpart of the build system, like Zephyr modules, DTS processing or Kconfig in process script mode with a single and uniform entry point.

A common `USAGE` mode can be supported for modules.

> But if you still believe that it should be supported, then it should be documented what happens when boards.cmake and shields.cmake is executed in script mode.

I'm mostly concerned on breaking existing functionality that someone is relying upon.

Reply:
I would appreciate a comment explaining what should happen when board.cmake is invoked in script mode for review purposes.

I can't check that something is doing what it is supposed to do if we don't say what it should do.

We didn't document this earlier because it was (and still is IMHO) an internal file.

Reply:
Removed the script mode support, see more here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#issuecomment-1004641968